File: hooks/block-nixos-rebuild-protected-changes.py
#!/usr/bin/env python3
"""PreToolUse hook: block nixos-rebuild when protected config lines are changed."""

import json
import re
import subprocess
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))
from nix_protected import check_config_diff, check_mkforce_override

NIX_CONFIG = Path.home() / "nix-config"


def get_diff(cwd: Path) -> str:
    parts = []
    for extra_args in [[], ["--cached"]]:
        result = subprocess.run(
            ["git", "diff", *extra_args],
            capture_output=True, text=True, cwd=cwd,
        )
        if result.returncode == 0:
            parts.append(result.stdout)
    return "\n".join(parts)


def main() -> None:
    data = json.load(sys.stdin)
    command = data.get("tool_input", {}).get("command", "")

    if not re.search(r"\bnixos-rebuild\b", command):
        return

    diff_text = get_diff(NIX_CONFIG)
    if not diff_text.strip():
        return

    reason = check_config_diff(diff_text)
    if not reason:
        reason = check_mkforce_override(diff_text)

    if reason:
        json.dump({"decision": "block", "reason": reason}, sys.stdout)


if __name__ == "__main__":
    main()

File: hooks-rs/src/hooks/block_nixos_rebuild_protected_changes.rs
use regex::Regex;

use crate::input::HookInput;
use crate::nix_protected::{check_config_diff, check_mkforce_override};
use crate::output::{block, pass};

pub fn run(input: &HookInput) {
    let command = &input.tool_input.command;
    if !Regex::new(r"\bnixos-rebuild\b").unwrap().is_match(command) {
        return;
    }

    let nix_config = home::home_dir().unwrap_or_default().join("nix-config");
    let mut parts = Vec::new();
    for extra in [&[][..], &["--cached".to_string()]] {
        let mut args = vec!["diff".to_string()];
        args.extend(extra.iter().cloned());
        let output = std::process::Command::new("git")
            .args(&args)
            .current_dir(&nix_config)
            .output();
        if let Ok(o) = output
            && o.status.success()
        {
            parts.push(String::from_utf8_lossy(&o.stdout).to_string());
        }
    }

    let diff_text = parts.join("\n");
    if diff_text.trim().is_empty() {
        return;
    }

    if let Some(reason) = check_config_diff(&diff_text) {
        block(&reason);
        return;
    }
    if let Some(reason) = check_mkforce_override(&diff_text) {
        block(&reason);
        return;
    }
    pass();
}
